### PR TITLE
add -vasm option that prints generated code per function

### PR DESCRIPTION
**Pull request by WalterBright**
This adds a disassembler built in to dmd that will present the assembly code generated for each function. It's based on the Digital Mars disassembler https://www.digitalmars.com/ctg/obj2asm.html

Why?

1. godbolt requires one to prepare a smallish sample code to upload. Too much work.
2. `obj2asm` works, but also prints out everything in the object file, which is usually Way Too Much Information. It's also too much work.
3. By making it ridiculously easy to see the code generated for each function, the user is encouraged to become familiar with what constructs generate what sorts of code. This will inevitably lead to the user writing better code.
4. Makes it easier to diagnose wrong code generated by the compiler, and easier to file bug reports for it.
5. User can easily tell if a critical function is being inlined or not.

Genesis

I originally wrote this disassembler around 1982-83, as a tool for developing the C compiler. I just needed it to work, and paid little attention to making it nice. Over time, it got extended to support 32 bit code, and then 64 bit code, and then all those extra instructions. It got extended to support all the object file formats dmd supports.

But it was a pity such a useful bit of code was imprisoned in obj2asm. For a long time, I wanted to liberate it. I finally got around to it. The first task was to disentangle the disassembler from all the file handling, fixups, and symbol table stuff. Then I translated it to D. The goal was to make disasm86.d be completely standalone, not dependent on any other D libraries.

So here it is, crudely bolted on to dmd, and Boost licensed. The code is still ugly, a legacy from 1982, but it works.

By making it more widely used, I expect it will be maintained better, and become more worthwhile to polish.

If you've ever wanted to add a disassembler to Emacs, your new debugger project, your own compiler explorer, IDE, etc., I've already done the boring part. Enjoy!

### CODE REVIEW

**Comment by WalterBright** [comment]
The meat of this is the `disassemble` function, and the disasm86.d file. The rest is just the boilerplate necessary to collect the option from the user and pass it along.

**Comment by RazvanN7** [comment]
I don't really see the benefit of adding 4k lines of code to the compiler to basically duplicate what `objdump` does. The fact that it spits out a lot of information is by no means a problem since you can "objdump ... | less" and search for the function that interests you.

File: src/dmd/cli.d [hunk 1]
```diff
@@ -704,6 +704,9 @@ dmd -cov -unittest myprog.d
             "verbose",
             `Enable verbose output for each compiler pass`,
         ),
+        Option("vasm",
```
Comment (dkorpel) [inline src/dmd/cli.d:707]: I found [a bug!](https://news.ycombinator.com/item?id=29570425)

File: src/dmd/backend/cgcod.d [hunk 2]
```diff
@@ -579,6 +583,10 @@ tryagain:
         }
         sfunc.Ssize = Offset(sfunc.Sseg) - funcoffset;    // size of function
 
+        //xyzzy disasm
```
Comment (dkorpel) [inline src/dmd/backend/cgcod.d:586]: What is "xyzzy"?

File: src/dmd/backend/cod3.d [hunk 3]
```diff
@@ -6784,26 +6785,24 @@ nomatch:
...
+ * later be sent along to objmod
  */
-
 private struct MiniCodeBuf
 {
 nothrow:
-    size_t index;
-    size_t offset;
+    uint index;
+    uint offset;
```
Comment (dkorpel) [inline src/dmd/backend/cod3.d:6796]: Why change `size_t` to `uint`?

File: src/dmd/backend/cod3.d [hunk 4]
```diff
@@ -6812,28 +6811,88 @@ nothrow:
...
+        }
+    }
 }
 
+/**************************
+ * Convert instructions to object code and write them to objmod.
+ * Params:
+ *      seg = code segment to write to, code starts at Offset(seg)
+ *      c = list of instructions to write
+ *      disasmBuf = if not null, then also write object code here
```
Comment (dkorpel) [inline src/dmd/backend/cod3.d:6889]: The looks fragile, the function now outputs in two places which can go out of sync. Imagine debugging code generation with this feature and finding out the `-vasm` assembly output is not actually what the compiler generated.

**Comment by WalterBright** [comment]
It's the same rationale as why build Ddoc and unittest into the compiler when 3rd party tools can do it better? Reducing friction can be a game changer.

Also, I did my best at making those 4000 lines of code not disruptive to the rest of the compiler.

File: src/dmd/backend/cgcod.d [hunk 5]
```diff
@@ -516,6 +517,9 @@ tryagain:
     }
     else
     {
+        __gshared Barray!ubyte disasmBuf;
```
Comment (dkorpel) [inline src/dmd/backend/cgcod.d:520]: Avoid global variables

**Review by dkorpel — DISMISSED** [review]

**Comment by WalterBright** [comment]
It's a feature I've long wished for, as I get tired of typing `obj2asm file.obj >log` all the time.

File: src/dmd/backend/cod3.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6796]: reduce stack size consumption

File: src/dmd/backend/cod3.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6889]: It was the easiest and least intrusive way to do it, by far.

File: src/dmd/backend/cgcod.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/cgcod.d:520]: It's done to avoid constantly allocating and reallocating the memory for it.

File: src/dmd/backend/cgcod.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/backend/cgcod.d:586]: a piece of debug code I forgot to remove

File: src/dmd/cli.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/cli.d:707]: I'm glad someone reads my sagacious postings on HN!

File: src/dmd/backend/cod3.d (hunk 3 above)
Comment (dkorpel) [reply src/dmd/backend/cod3.d:6796]: A PR should do one thing, not mix adding a feature and doing optimizations

File: src/dmd/backend/cgcod.d (hunk 5 above)
Comment (dkorpel) [reply src/dmd/backend/cgcod.d:520]: Then let the caller initialize it

File: src/dmd/backend/cod3.d (hunk 4 above)
Comment (dkorpel) [reply src/dmd/backend/cod3.d:6889]: It's not like you're outputting backend IR, disassembly is not something DMD specific, I don't see why it needs to be interwoven with the backend.

File: src/dmd/cli.d (hunk 1 above)
Comment (dkorpel) [reply src/dmd/cli.d:707]: I love your HN posts! 

But to elaborate, when I asked at DConf online what to do about the ever growing list of command line options in DMD, you answered "I got to say 'no' more often". Can you say 'no' to this switch and create a shell script that calls `dmd + obj2asm` instead?

File: src/dmd/mars.d [hunk 6]
```diff
@@ -1751,6 +1751,8 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
             params.verbose = true;
         else if (arg == "-vcg-ast")
             params.vcg_ast = true;
+        else if (arg == "-vasm") // https://dlang.org/dmd.html#switch-vasm
+            params.vasm = true;
```
Comment (ibuclaw) [inline src/dmd/mars.d:1755]: Why not put this in DMDparams, given that it is not a front-end option?

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/cli.d (hunk 1 above)
Comment (maxhaton) [reply src/dmd/cli.d:707]: The flag could also be just be the compiler running that command for you. 

By far best (e.g. look at what Intel C++ can do) textual emission of assembly on Compiler Explorer is generated from the last IR (i.e. just before actually turning it into actual executable bytes) in the compiler rather than actually dissassembling the binary. dmd on compiler explorer does the latter which is why the output is full of disssssembled padding for example.

**Comment by maxhaton** [comment]
There seems to be 3 reimplementations of `puts` in this new module

**Comment by maxhaton** [comment]
https://blog.trailofbits.com/2019/10/31/destroying-x86_64-instruction-decoders-with-differential-fuzzing/ 

Also it's worth noting that one can get X86 decoding very wrong, so this module would likely benefit from use of `unittest`. Can these actually work considering the backend is betterC.

File: src/dmd/backend/cod3.d (hunk 4 above)
Comment (maxhaton) [reply src/dmd/backend/cod3.d:6889]: > It's not like you're outputting backend IR, disassembly is not something DMD specific, I don't see why it needs to be interwoven with the backend.

GCC and LLVM both already do this in such a way that I can't see anyone using this if it were not specific to dmd, especially if consider they'd be adding a fairly large blob of code just to support X86 (but not the latest instructions) but not any of their other targets.

**Comment by maxhaton** [comment]
For this to actually combat the issue mentioned above (i.e. objdump having too much output, godbolt requiring a reduced sample - which I note is actually quite easy (95% of the time) to do if you understand your code) then it should probably be able to take a glob or something similar otherwise presumably it too will also output huge amounts of output. 

Our binaries at Symmetry would probably generate a a gigabyte of text if everything was emitted.

**Comment by WalterBright** [comment]
> Also it's worth noting that one can get X86 decoding very wrong,

I'm very well aware of that! Keep in mind that I used this disassembler for 40 years (in the guise of obj2asm) to debug the code generator. I built it to debug the code generator with. It has been very well tested, although it does lack comprehensive unittests. One reason to include it in dmd is to add usage so it will be even better tested. It's dmd usage is also benign in that its output is informational only.

> so this module would likely benefit from use of unittest. Can these actually work considering the backend is betterC.

It has unittests, and it can be unittested as a standalone program (which is how I developed it). I plan on adding it to dub as a standalone module.

**Comment by WalterBright** [comment]
> Our binaries at Symmetry would probably generate a a gigabyte of text if everything was emitted.

I'm sure it would. Running obj2asm on such large files can also be an exercise in frustration. Where I find this most useful, however, is when looking at a reduced test case. -vasm means the output I'm interested in fits on a single screen without all the extra stuff that obj2asm emits.

File: src/dmd/mars.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/mars.d:1755]: I didn't know about DMDparams.

File: src/dmd/backend/cgcod.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/cgcod.d:520]: It's have to go back multiple levels.

File: src/dmd/backend/cod3.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6889]: disasm86 supports the latest instructions, just like the inline assembler does. I agree this is a dmd backend specific feature, not a D feature, and what ldc and gdc decide to do about it is entirely up to them.

File: src/dmd/cli.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/cli.d:707]: Doing it in dmd before generating the actual executable bytes would have been a large project that would have been useless for any other purpose, hence would have to be separately maintained. This was a small project, and supporting the complete instruction set is something I regularly do for it anyway, so it is not extra work to keep it up to date.

It's value is also increased as it is designed to be usable for any program that might benefit from a disassembler.

File: src/dmd/backend/cod3.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6889]: > What if you disassemble the .obj file around the time it would be written to disk?

Because of the complexity of the OMF file format, again, this way is the easiest and least intrusive way to do it, by far. I also don't have to write a separate decoder for each supported object file format.

**Comment by WalterBright** [comment]
> There seems to be 3 reimplementations of puts in this new module

There are 2, and they are nested functions that need access to their local context.

File: src/dmd/backend/cod3.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6796]: 2 lines out of 4000 is not a problem.

Comment (dkorpel) [reply src/dmd/backend/cod3.d:6796]: So if this were to introduce a regression and you would git bisect, you would debug a diff of 4000 lines. This is exactly the thing you are vehemently against in other people's PRs.

File: src/dmd/backend/cgcod.d (hunk 5 above)
Comment (dkorpel) [reply src/dmd/backend/cgcod.d:520]: I thought you wanted to move dmd in a `@safe` and `pure` direction. I understand legacy code using `__gshared`, but why write new code using it?

File: src/dmd/backend/disasm86.d [hunk 7]
```diff
@@ -0,0 +1,3758 @@
...
+                                else
+                                    p = (rex ? byteregrex[rm] : bytereg[rm]);
+                                if (rex & REX_W)
+                                    p = (opcode & 1 || r32) ? rreg[rm] : byteregrex[rm];
+                            }
+                            break;
+                    }
+                    break;
+                default:
+                    assert(0);
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]: `final switch`?

File: src/dmd/backend/disasm86.d [hunk 8]
```diff
@@ -0,0 +1,3758 @@
...
+    __gshared const(char)*[8] op7 =
+    [   "fnild","fnisttp","fnist","fnistp","fnbld","fnild","fnbst","fnistp"];
+    __gshared const(char)*[7] op7b = [ "fnfree","fnxch","fnstp","fnstp","fnstsw","fnucomip","fncomip" ];
+    __gshared const(char)*[4] opD9 = [ "fnldenv","fnldcw","fnstenv","fnstcw" ];
+    __gshared const(char)*[6] opDDb = ["fnfree","fnxch","fnst","fnstp","fnucom","fnucomp"];
+    __gshared const(char)*[8] opDDa = ["fnld","fnisttp","fnst","fnstp",
+            "fnrstor","reserved","fnsave","fnstsw"];
+    __gshared const(char)*[5] opDBa = [ "fneni","fndisi","fnclex","fninit","fnsetpm" ];
+    __gshared const(char)* ST = "ST";
+    __gshared const(char)* STI = "ST(i)";
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]: Make this `static immutable`.

File: src/dmd/backend/disasm86.d [hunk 9]
```diff
@@ -0,0 +1,3758 @@
...
+                if ((modrgrm & 0xC0) == 0xC0)
+                {   p1 = (reg <= 6) ? op7b[reg] : reserved;
+                    if (reg == 4)
+                        p2 = "AX";
+                    else
+                    {   if (reg == 5)
+                            mfp = "";
+                        if ((modrgrm & 7) != 1)
+                        {
+                            strcpy(sti.ptr, STI);
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]: Try to replace the rest of the `strcpy` calls with equivalent slice syntax.

File: src/dmd/backend/disasm86.d [hunk 10]
```diff
@@ -0,0 +1,3758 @@
...
+    const(char)* sep;
+    const(char)* s2;
+    const(char)* s3;
+    char[BUFMAX] buf = void;
+    enum MOV = "mov";
+    enum XCHG = "xchg";
+    enum IMUL = "imul";
+    enum SHRD = "shrd";
+    enum SHLD = "shld";
+    __gshared const char*[12] astring =
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]: ditto

File: src/dmd/backend/disasm86.d [hunk 11]
```diff
@@ -0,0 +1,3758 @@
...
+    assert(siz > prefixsize);
+    siz -= prefixsize;
+
+    put('\t');
+
+    uint opcode,reg = 0;
+    ubyte rex;
+    int i,o3;
+    const(char)* p1, p2, p3, p4;
+    char[80] p0;
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]: Since you are doing `p0[0]='\0';` and `buf` is void initialized, why not do it to the rest of the buffers to remain consistent? I also suggest adding those assignments together to be easier to read.

File: src/dmd/backend/disasm86.d [hunk 12]
```diff
@@ -0,0 +1,3758 @@
...
+    char[5] p1buf;
+    const(char)* sep;
+    const(char)* s2;
+    const(char)* s3;
+    char[BUFMAX] buf = void;
+    enum MOV = "mov";
```
Comment (ljmf00) [inline src/dmd/backend/disasm86.d]:
```suggestion
    static immutable MOV = "mov";
```
Avoid using `enum` for arrays, appending to it might not generate a new compile-time string and instead calls to append it at runtime.

**Review by ljmf00 — DISMISSED** [review]
Rough review. Try to remove `__gshared const char*` and annotate functions to be `pure`. I understand that this code was ported from C but avoiding `strlen`, `strcpy` and other similar calls can improve debugging this in the future with bound checks, some of them possibly discoverable at compile-time.

I would also suggest adding more tests since this new file is only 19% of coverage. At least add a shell test to the test suite to prove `-vasm` functionality.

**Comment by WalterBright** [comment]
@ljmf00 These are all good suggestions, and will improve the code. But consider that this code was translated from long time working C code. In my extensive experience doing translations, the initial code needs to *avoid* doing refactorings, and concentrate on having a strong correspondence between the original and the translated code. This is because, inevitably, bugs are introduced in the translation. Having the strong correspondence makes it relatively easy to discover where things went wrong.

And, indeed, in preparing this I *did* introduce problems and was able to track them down quickly and easily.

The idea here is to check in the working code, warts and all, and *then* proceed with refactorings. Then, if things break, `git bisect` has a chance of determining if things went awry in the translation, or in the refactoring.

I don't even like to put a lot of refactorings into one PR, spreading them out into multiple ones.

This process is exactly how the backend was translated from C to D. I refactor it from time to time to remove the old C cruft and modernize it, as time permits. It's an ongoing process. But by being incremental, I have saved myself a *great* deal of grief.

File: src/dmd/backend/cod3.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d:6796]: 4000 lines? Not at all. This code is not part of the disassembler, I was very careful in isolating the disassembler from internals to the rest of the backend. The changes to the MiniCodeBuf are minimal and easily reviewed.

Secondly, I point out repeatedly that rules for doing things often conflict, and one must always exercise *good judgement*. My complaints were about PRs that would have one line of bug fix and 400 lines of refactorings. Yes, this sort of thing happened repeatedly. If you check my comment history, I don't complain about a line here and there being improved in the context of a larger change.

Thirdly, changing the type got rid of a couple ugly casts.

File: src/dmd/backend/cgcod.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/cgcod.d:520]: The idea is to get things to work with minimal changes to the code base. Refactorings come later. Figuring out how to remove globals is a case-by-case thing, and unnecessarily adds complexity to this PR.

**Comment by ljmf00** [comment]
> @ljmf00 These are all good suggestions, and will improve the code. But consider that this code was translated from long time working C code. In my extensive experience doing translations, the initial code needs to _avoid_ doing refactorings, and concentrate on having a strong correspondence between the original and the translated code. This is because, inevitably, bugs are introduced in the translation. Having the strong correspondence makes it relatively easy to discover where things went wrong.
> 
> And, indeed, in preparing this I _did_ introduce problems and was able to track them down quickly and easily.
> 
> The idea here is to check in the working code, warts and all, and _then_ proceed with refactorings. Then, if things break, `git bisect` has a chance of determining if things went awry in the translation, or in the refactoring.
> 
> I don't even like to put a lot of refactorings into one PR, spreading them out into multiple ones.
> 
> This process is exactly how the backend was translated from C to D. I refactor it from time to time to remove the old C cruft and modernize it, as time permits. It's an ongoing process. But by being incremental, I have saved myself a _great_ deal of grief.

Thanks for your reply. I understand some of your points on this, but I think introducing ~4k of LOC without much review is generally not a good idea, especially dealing with backend code. I'm not against this addition at all if it doesn't produce side effects for the rest of the backend logic. I think we can still iteratively add patches on top of the initial change in the same PR and keep them in the history, so that `git bisect` works just fine.

**Comment by WalterBright** [comment]
> introducing ~4k of LOC without much review

Generally speaking, you're right. But consider that this code simply prints information to the screen. It has no connection to the rest of the back end other than filling a buffer with text. It's very low risk, even if it is completely broken.

> add patches on top of the initial change in the same PR and keep them in the history, so that git bisect works just fine.

I don't care for that approach. For one thing, the individual commits are not forced to actually work, so the git bisect will produce false positives. For another, there's never going to be usage feedback on an uncommitted PR. Third, this is a way to avoid making a decision. Fourth, there's no such thing as ever being done with a piece of software.

This is a very low risk PR, despite the number of lines.

**Comment by ljmf00** [comment]
> > introducing ~4k of LOC without much review
> 
> Generally speaking, you're right. But consider that this code simply prints information to the screen. It has no connection to the rest of the back end other than filling a buffer with text. It's very low risk, even if it is completely broken.

I can't agree much with that, since this is going to be publicly available for the end-user, as a feature. At least we should test it. According to my Sourcegraph plugin, the introduced file have 19% of coverage and according to [Codecov reports](https://app.codecov.io/gh/dlang/dmd/branch/disasm), this will reduce around 2.53% of coverage on the backend code, and we should take care of coverage, especially the backend, which is the least covered code of the entire codebase. What I mean by that is, even without entanglement with other backend code, as a whole, the backend is considered less reliable.

> > add patches on top of the initial change in the same PR and keep them in the history, so that git bisect works just fine.
> 
> I don't care for that approach. For one thing, the individual commits are not forced to actually work, so the git bisect will produce false positives.

If you are adding commits that are not supposed to work, or at least don't care about its compiling state, then maybe you are doing commits wrongly. Optimally a commit should compile on its own, although, in practice, that doesn't apply if we have external dependencies. If the patches are added now or later it doesn't matter for `git bisect`, we just need to be careful of amending/fixup the right commit.

That being said, I understand that this approach can introduce some "complexity" in the PR and for the committer, if not familiar with `git commit --fixup=` workflow.

What I suggest, if you don't want to refactor this code now, would be to make this option not publicly exposed until this feature is considered stable enough and add some integration testing to prove the basic functionality of the introduced feature, eventually add refactoring PRs on top of this and then publicly expose it and add a changelog entry. We can get a rough list of things to tick off:

- [ ]  Use `static immutable` instead of `enum` for arrays
- [ ] Avoid usage of `__gshared` as much as possible for immutable content or for creation of local static variables making the method impure and not thread-safe
- [ ] Remove usage of unsafe and unchecked calls like `memcpy`, `strcpy`, `strcat`, ... and use slice syntax instead
- [ ] Annotate functions with `@safe`, `pure`,... attributes

We should also list the functionality this is introducing such as what opcodes and x86 extensions this disassembler supports (16-bit, 32-bit, 64-bit x86 opcodes, x87, XMM, and other vector extensions) and add tests for it. We should also add tests on broken opcodes and define failing behaviours.

File: src/dmd/backend/cgcod.d [hunk 13]
```diff
@@ -3467,4 +3474,32 @@ void andregcon(con_t *pregconsave)
     regcon.cse.mops &= regcon.cse.mval;
 }
 
+
+/**********************************************
+ * Disassemble the code instruction bytes
+ * Params:
+ *    code = array of instruction bytes
+ */
+@trusted
```
Comment (ljmf00) [inline src/dmd/backend/cgcod.d:3483]: I think this is concerning. This is basically `@trusted` 4k of lines?

Comment (WalterBright) [reply src/dmd/backend/cgcod.d:3483]: On the backend, I added @trusted to everything, as a first step towards using @safe. It's much harder to find the @safe code without doing that.

**Comment by WalterBright** [comment]
If it's not publicly exposed, nobody can use it.

BTW, adding tests for every opcode will be several thousand lines. It's not impractical, and should be done, but this code is already heavily tested (it's been in constant use for 40 years now), just not unittested. It's also simply not critical at the moment.

**Comment by WalterBright** [comment]
P.S. I'd also like to see how useful it is before investing a great deal more time into it.

File: src/dmd/backend/disasm86.d [hunk 14]
```diff
@@ -0,0 +1,3758 @@
+/*********************************************************
+ * X86 disassembler. Can disassemble 16, 32, and 64 bit code. Includes
+ * x87 FPU instructions and vector instructions.
+ *
+ * Copyright:   Copyright (C) 1982-1998 by Symantec
+ *              Copyright (C) 2000-2021 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
```
Comment (ibuclaw) [inline src/dmd/backend/disasm86.d]:
```suggestion
 * Authors:     $(LINK2 https://www.digitalmars.com, Walter Bright)
 * License:     $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
```

**Comment by ibuclaw** [comment]
> P.S. I'd also like to see how useful it is before investing a great deal more time into it.

It seems that one compromise would be to have it as a dmd backend debug option then (`--a`?) just so that it is there for initial testing before making it a public option, because otherwise you'd have to deal with all the baggage and churn that comes with exposing a feature to the front-end CLI (just look at how disastrous `-target=` was and still is).

File: src/dmd/backend/disasm86.d (hunk 14 above)
Comment (WalterBright) [reply src/dmd/backend/disasm86.d]: done

**Comment by WalterBright** [comment]
> all the baggage and churn

`-target` influences the compiler semantic behavior. `-vasm` does not, so there shouldn't be baggage and churn.

**Comment by WalterBright** [comment]
> error: RPC failed

It would really be nice to fix the testers. Well over half of the time one or more of the tests randomly fail from environmental issues. It seems to get steadily worse.

The fix is to detect environmental errors, pause for a few seconds, and retry instead of just falling over.

**Comment by ljmf00** [comment]
> If it's not publicly exposed, nobody can use it.
> 
> BTW, adding tests for every opcode will be several thousand lines. It's not impractical, and should be done, but this code is already heavily tested (it's been in constant use for 40 years now), just not unittested. It's also simply not critical at the moment.

At least add a simple test case using `-vasm` and check its stdout directly to prove minimum functionality.

**Comment by maxhaton** [comment]
If this was an asm naked function then that's a good idea, I don't think we
should be testing dmd's code generation directly.

**Comment by WalterBright** [comment]
-vasm will of course emit output that is highly dependent on the target and compiler switches. It's more of a nuisance testing it against fixed files than a help. We don't do it for other types of helpful things the compiler emits for the same reason.

**Comment by ljmf00** [comment]
> -vasm will of course emit output that is highly dependent on the target and compiler switches. It's more of a nuisance testing it against fixed files than a help. We don't do it for other types of helpful things the compiler emits for the same reason.

As mentioned by @maxhaton , is this consistently achievable with nacked inline asm?

**Comment by WalterBright** [comment]
> As mentioned by @maxhaton , is this consistently achievable with nacked inline asm?

No. The final register assignments, for example, are highly dependent on everything else.

**Comment by RazvanN7** [comment]
So, are we gonna pull the trigger on this one?

**Comment by ljmf00** [comment]
> > As mentioned by @maxhaton , is this consistently achievable with nacked inline asm?
> 
> No. The final register assignments, for example, are highly dependent on everything else.

This might need some extra changes on the shell testsuite checker, but I think this is possible to be tested. I assume the body doesn't get affected. e.g. https://godbolt.org/z/cs6eaK4Tx . We can create a test to check from the beggining of the function until `ret`. On LLVM, we have `CHECK` and `CHECK_NEXT` and the last one need to be followed by the former.

Also, why `add [EAX],AL` next to a `ret` is needed when the function only contains inline asm and is naked? Seems always unreachable to me in those conditions.

> So, are we gonna pull the trigger on this one?

From my side and related to my review, I don't want to be a blocker. I think this is a reasonable addition as it doesn't affect other code/behaviour. I just expected some more coverage on such big addition. Given the reasons, I think this shouldn't be available as a "functional flag" on the compiler until we have a considerable testsuite and refactor the unsafe/unchecked calls mentioned.

**Comment by maxhaton** [comment]
add [EAX],AL  assembled is 00 00 i.e. it's padding. You see this on godbolt
because it has to disassemble the resulting binary so you get the warts and
all.

This should probably be changed to be a NOP since the frontend of the
processor might try and do something with this (at very least it'll use
more power than a NOP, it depends how clever the CPU is when it sees a
ret), but dmd is slow enough that anyone who cares about that wouldn't be
using it at that point so who cares.

Anyway we can probably merge this, I will get round to fuzzing it at some
point (i.e. It would be a pleasant surprise if this was competitive to xed
and capstone for instance), my only gripe is whether this actually does
what it supposedly does better than merely objumping the binary after
compilation - if you could provide a glob of the functions you want to
disassemble then that would be a better experience than naive objdump, but
if it just spews out every single MapResult.empty from a large binary when
you wanted to look at FooBar then it is harder to motivate such a huge
addition to the backend. That being said I favour on the side of merging
and finding out, it can't hurt that much.

**Comment by WalterBright** [comment]
And it will save *me* time when I'm debugging the code generator, or checking to make sure the expected code is actually generated.

**Review by RazvanN7 — APPROVED** [review]

**Comment by ljmf00** [comment]
Since this is going to be public, aren't we going to add a simple changelog entry?

**Comment by WalterBright** [comment]
https://github.com/dlang/dmd/pull/13500
